More robust init

benoit-pierre opened the pull request:
Fix #273.

* use blanket except clauses so Plover does not crash if something goes wrong during the init.
* use the logger for reporting exceptions.
* use a custom formatter for GUI log handlers: don't show traceback on exception, since the messages are usually limited in length, and the information is not really useful to most users (the full traceback is still logged to the console and the logfile).

morinted commented:
- Plover doesn't complain when opening multiple times now.
- When I open without assistive devices enabled, I'm presented with the error and the configuration dialogue shows up. So then I select "Tréal" from the machine list, and hit "save". Then, the user interface shows "Gemini PR" while the console is talking about the Tréal, so I try to close Plover, and then the user interface turns into a shell and no longer responds
<img width="325" alt="screen shot 2016-03-01 at 3 58 38 pm" src="https://cloud.githubusercontent.com/assets/5840970/13441675/7b05d526-dfc6-11e5-8842-e465d57c87d8.png">

Log is mostly empty:

```
2016-03-01 15:56:41,632 ERROR: updating engine configuration failed
Traceback (most recent call last):
   […]
    raise Exception("Enable access for assistive devices.")
Exception: Enable access for assistive devices.
2016-03-01 15:56:56,463 INFO: Treal device not found: open failed
2016-03-01 15:56:56,464 WARNING: Treal is not connected
2016-03-01 15:56:56,478 INFO: Treal device not found: open failed
2016-03-01 15:56:56,478 WARNING: Treal is not connected
2016-03-01 15:56:56,481 INFO: set_stroke_filename(/Users/ted/Library/Application Support/plover/strokes.log)
2016-03-01 15:56:56,481 INFO: enable_stroke_logging(True)
2016-03-01 15:56:56,481 INFO: enable_translation_logging(True)
```

benoit-pierre commented:
> Plover doesn't complain when opening multiple times now.

I can't reproduce that (both on Linux and with a Windows 10 VM).

morinted commented:
Ah, yeah, seems that I was mistaken on that one.

benoit-pierre commented:
> When I open without assistive devices enabled, I'm presented with the error and the configuration dialogue shows up. So then I select "Tréal" from the machine list, and hit "save". Then, the user interface shows "Gemini PR" while the console is talking about the Tréal, so I try to close Plover, and then the user interface turns into a shell and no longer responds 

I tried simulating that by adding a call to `raise Exception("Enable access for assistive devices.")` in `oslayer/xkeyboardcontrol.py`, and switching to "Tréal" when the configuration dialog shows up and it works.

benoit-pierre commented:
You're log also does not match the errors you're describing: 
```
2016-03-01 15:56:56,478 INFO: Treal device not found: open failed
2016-03-01 15:56:56,478 WARNING: Treal is not connected
```

morinted commented:
Well, in that case, the Treal wasn't connected. The log was complaining about the Treal, but the UI is showing Gemini PR.

morinted commented:
Here's a little more detail.

- Open Plover: engine initialization failed:
<img width="495" alt="screen shot 2016-03-01 at 4 08 54 pm" src="https://cloud.githubusercontent.com/assets/5840970/13441963/fd9ff0e2-dfc7-11e5-8456-2a97923a06fa.png">
- Switch to Treal as machine, hit save...
<img width="491" alt="screen shot 2016-03-01 at 4 09 07 pm" src="https://cloud.githubusercontent.com/assets/5840970/13441962/fd9d464e-dfc7-11e5-88b1-e37dd1814c7f.png">
- Resulting state, pressing keys on the Treal has no effect:
<img width="813" alt="screen shot 2016-03-01 at 4 09 17 pm" src="https://cloud.githubusercontent.com/assets/5840970/13441964/fd9ff8ee-dfc7-11e5-8488-a7e62b0b3731.png">

benoit-pierre commented:
If you start with the config set to `Machine1`, and there's an error during init, the configuration dialog is show, and if you then switch to `Machine2`, but the configuration update fails too (which can happen when retrieving the machine options, not just for the machine init), then the interface still shows `Machine1`. Is that what is happening?

morinted commented:
I'm starting with Keyboard, switching to Tréal, and then it's giving me this state. If I force close Plover and restart it, the Tréal is the selected device.

benoit-pierre commented:
OK, so on OSX with WX 3.0, `wx.CallAfter` cease to work after showing the modal configuration dialog in the GUI init loop... I removed said loop and fixed the logic in `update_engine`.

morinted commented:
I ran through this on Windows and OSX… It is indeed much more stable. I can break the dictionary configuration, break the machine, it all resumes correctly. It's very satisfying
